ogre2/src/Ogre2ParticleEmitter.cc

```diff
@@ -14,17 +14,380 @@
  * limitations under the License.
  *
  */
+
+// Note this include is placed in the src file because
+// otherwise ogre produces compile errors
+#include <Hlms/Pbs/OgreHlmsPbsDatablock.h>
+#include <Hlms/Unlit/OgreHlmsUnlitDatablock.h>
+
+#include "ignition/rendering/ogre2/Ogre2Conversions.hh"
+#include "ignition/rendering/ogre2/Ogre2Includes.hh"
+#include "ignition/rendering/ogre2/Ogre2Material.hh"
 #include "ignition/rendering/ogre2/Ogre2ParticleEmitter.hh"
+#include "ignition/rendering/ogre2/Ogre2RenderEngine.hh"
+#include "ignition/rendering/ogre2/Ogre2Scene.hh"
 
 using namespace ignition;
 using namespace rendering;
 
+class ignition::rendering::Ogre2ParticleEmitterPrivate
+{
+  /// \brief Internal material name.
+  public: const std::string kMaterialName = "__particle_emitter_material__";
+
+  /// \brief Ogre particle system.
+  public: Ogre::ParticleSystem *ps = nullptr;
+
+  /// \brief Ogre particle emitter.
+  public: Ogre::ParticleEmitter *emitter = nullptr;
+
+  // \brief Ogre colour image affector.
+  public: Ogre::ParticleAffector *colourImageAffector = nullptr;
+
+  // \brief Ogre colour interpolator affector.
+  public: Ogre::ParticleAffector *colourInterpolatorAffector = nullptr;
+
+  /// \brief Ogre scaler affector.
+  public: Ogre::ParticleAffector *scalerAffector = nullptr;
+
+  /// \brief Pointer to the material datablock.
+  public: Ogre::HlmsUnlitDatablock *ogreDatablock = nullptr;
+};
+
 //////////////////////////////////////////////////
 Ogre2ParticleEmitter::Ogre2ParticleEmitter()
+    : dataPtr(new Ogre2ParticleEmitterPrivate)
 {
 }
 
 //////////////////////////////////////////////////
 Ogre2ParticleEmitter::~Ogre2ParticleEmitter()
 {
 }
+
+//////////////////////////////////////////////////
+void Ogre2ParticleEmitter::Ogre2ParticleEmitter::SetType(
+    const EmitterType _type)
+{
+  switch (_type)
+  {
+    case EmitterType::EM_POINT:
+    {
+      this->dataPtr->ps->removeAllEmitters();
```

Comment:
I think with the remove call here, calling `SetType` will remove existing emitter settings, e.g. size, duration, rate, etc?

Reply:
Oh that was a bug, good catch. Fixed in 571a08f. 

In order to make it work, we now keep an array of emitters, one of each supported type. We apply all emitter attributes to all each emitter but only one is visible at a time. When the function `SetType()` is called we switch to the appropriate emitter.